File: cpp/hot100/47.从前序与中序遍历序列构造二叉树.cpp
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int n=preorder.size();
        unordered_map<int,int> index;
        for(int i=0;i<n;i++){
            index[inorder[i]]=i;
        }
        auto dfs=[&](this auto&& dfs,int pre_l,int pre_r,int in_l,int in_r)->TreeNode*{
            if(pre_l==pre_r) return nullptr;
            int left_size=index[preorder[pre_l]]-in_l;
            TreeNode *left=dfs(pre_l+1,pre_l+1+left_size,in_l,in_l+left_size);
            TreeNode *right=dfs(pre_l+1+left_size,pre_r,in_l+left_size+1,in_r);
            return new TreeNode(preorder[pre_l],left,right);
        };
        return dfs(0,n,0,n);
    }
};

File: python/hot100/47.从前序与中序遍历序列构造二叉树.py
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        index={x: i for i,x in enumerate(inorder)}
        def dfs(pre_l,pre_r,in_l,in_r)->Optional[TreeNode]:
            if pre_l==pre_r:
                return None
            left_size=index[preorder[pre_l]]-in_l
            left=dfs(pre_l+1,pre_l+left_size+1,in_l,in_l+left_size)
            right=dfs(pre_l+left_size+1,pre_r,in_l+left_size+1,in_r)
            return TreeNode(preorder[pre_l],left,right)
        return dfs(0,len(preorder),0,len(preorder))